src/transformers/models/auto/tokenization_auto.py

```diff
@@ -29,315 +29,183 @@
+            ),
+        ),
+        ("ibert", ("RobertaTokenizer", "RobertaTokenizerFast" if is_tokenizers_available() else None)),
+        ("wav2vec2", ("Wav2Vec2CTCTokenizer", None)),
+        ("hubert", ("Wav2Vec2CTCTokenizer", None)),
+        ("gpt_neo", ("GPT2Tokenizer", "GPT2TokenizerFast" if is_tokenizers_available() else None)),
+        ("luke", ("LukeTokenizer", None)),
+        ("bigbird_pegasus", ("PegasusTokenizer", "PegasusTokenizerFast" if is_tokenizers_available() else None)),
+        ("canine", ("CanineTokenizer", None)),
+        ("bertweet", ("BertweetTokenizer", None)),
+        ("bert-japanese", ("BertJapaneseTokenizer", None)),
+        ("byt5", ("ByT5Tokenizer", None)),
+        (
+            "cpm",
+            (
+                "CpmTokenizer" if is_sentencepiece_available() else None,
+                "CpmTokenizerFast" if is_tokenizers_available() else None,
+            ),
+        ),
+        ("herbert", ("HerbertTokenizer", "HerbertTokenizerFast" if is_tokenizers_available() else None)),
+        ("phobert", ("PhobertTokenizer", None)),
+        (
+            "barthez",
+            (
+                "BarthezTokenizer" if is_sentencepiece_available() else None,
+                "BarthezTokenizerFast" if is_tokenizers_available() else None,
+            ),
+        ),
+        (
+            "mbart50",
+            (
+                "MBart50Tokenizer" if is_sentencepiece_available() else None,
+                "MBart50TokenizerFast" if is_tokenizers_available() else None,
+            ),
+        ),
     ]
 )
 
-# For tokenizers which are not directly mapped from a config
-NO_CONFIG_TOKENIZER = [
-    BertJapaneseTokenizer,
-    BertweetTokenizer,
-    ByT5Tokenizer,
-    CpmTokenizer,
-    CpmTokenizerFast,
-    HerbertTokenizer,
-    HerbertTokenizerFast,
-    PhobertTokenizer,
-    BarthezTokenizer,
-    BarthezTokenizerFast,
-    MBart50Tokenizer,
-    MBart50TokenizerFast,
-    PreTrainedTokenizerFast,
-]
-
-
-SLOW_TOKENIZER_MAPPING = {
-    k: (v[0] if v[0] is not None else v[1])
-    for k, v in TOKENIZER_MAPPING.items()
-    if (v[0] is not None or v[1] is not None)
-}
+TOKENIZER_MAPPING = _LazyAutoMapping(CONFIG_MAPPING_NAMES, TOKENIZER_MAPPING_NAMES)
+
+CONFIG_TO_TYPE = {v: k for k, v in CONFIG_MAPPING_NAMES.items()}
 
 
 def tokenizer_class_from_name(class_name: str):
-    all_tokenizer_classes = (
-        [v[0] for v in TOKENIZER_MAPPING.values() if v[0] is not None]
-        + [v[1] for v in TOKENIZER_MAPPING.values() if v[1] is not None]
-        + [v for v in NO_CONFIG_TOKENIZER if v is not None]
-    )
-    for c in all_tokenizer_classes:
-        if c.__name__ == class_name:
-            return c
+    if class_name == "PreTrainedTokenizerFast":
+        return PreTrainedTokenizerFast
+
+    for module_name, tokenizers in TOKENIZER_MAPPING_NAMES.items():
+        if class_name in tokenizers:
+            break
+
+    module = importlib.import_module(f".{module_name}", "transformers.models")
```

Comment:
This breaks for any model name with `-` in it. e.g. `xlm-roberta`,

For example:

```

Traceback (most recent call last):
  File "/mnt/nvme1/code/huggingface/transformers-master/examples/pytorch/language-modeling/run_mlm.py", line 550, in <module>
    main()
  File "/mnt/nvme1/code/huggingface/transformers-master/examples/pytorch/language-modeling/run_mlm.py", line 337, in main
    tokenizer = AutoTokenizer.from_pretrained(model_args.model_name_or_path, **tokenizer_kwargs)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/models/auto/tokenization_auto.py", line 432, in from_pretrained
    tokenizer_class = tokenizer_class_from_name(tokenizer_class_candidate)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/models/auto/tokenization_auto.py", line 226, in tokenizer_class_from_name
    module = importlib.import_module(f".{module_name}", "transformers.models")
  File "/home/stas/anaconda3/envs/py38-pt19/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'transformers.models.xlm-roberta'
```

as you can see it tries to import "transformers.models.xlm-roberta", to reproduce:

```
PYTHONPATH=src python examples/pytorch/language-modeling/run_mlm.py --train_file tests/fixtures/sample_text.txt --model_name_or_path hf-internal-testing/tiny-xlm-roberta --do_train --max_train_samples 4 --per_device_train_batch_size 2 --num_train_epochs 1 --fp16 --report_to none --overwrite_output_dir --output_dir output_dir --save_steps 1
```

```
# module_name, tokenizers debug print:
xlm-roberta ('XLMRobertaTokenizer', 'XLMRobertaTokenizerFast')
```

Should it do:
```
 module = importlib.import_module(f".{module_name.replace('-', '_')}", "transformers.models")
```

Oddly enough I don't get this problem if I run `xlm-roberta-base`, so this is an edge case.

Probably should include this tiny model as a test as it triggers this bug.

@LysandreJik, @sgugger

Reply:
You can see the failure on CI:

https://github.com/huggingface/transformers/runs/3406045740?check_suite_focus=true

Reply:
Proposed fix: https://github.com/huggingface/transformers/pull/13251